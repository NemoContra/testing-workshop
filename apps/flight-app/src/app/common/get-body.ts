import { inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';

export const getHTMLBodyElement = (): HTMLBodyElement =>
  inject(DOCUMENT).body as HTMLBodyElement;
