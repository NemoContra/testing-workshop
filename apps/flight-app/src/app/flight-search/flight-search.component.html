<div class="card">
  <div class="header">
    <h2 class="title">Flight Search</h2>
  </div>
  <div class="content">
    <form [formGroup]="searchFlightsFormGroup">
      <div class="form-group">
        <label>From:</label>
        <input formControlName="from" class="form-control" />
      </div>
      <div class="form-group">
        <label>To:</label>
        <input formControlName="to" name="to" class="form-control" />
      </div>

      <div *ngIf="flights$ | async as flights">
        {{ flights.length }} flights found!

        <div class="flight-cards">
          <flight-card
            *ngFor="let flight of flights"
            [flight]="flight"
          ></flight-card>
        </div>
      </div>
    </form>
  </div>
</div>
