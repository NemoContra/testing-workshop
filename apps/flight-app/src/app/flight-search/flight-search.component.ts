import { ChangeDetectionStrategy, Component, OnDestroy } from '@angular/core';
import {
  debounceTime,
  distinctUntilChanged,
  Observable,
  Subject,
  switchMap,
  takeUntil,
} from 'rxjs';
import {
  Flight,
  SearchFlightsQueryParams,
} from '@testing-workshop/shared/util/api-interfaces';
import { FlightService } from '../services/flight.service';
import { FormControl, FormGroup } from '@angular/forms';

@Component({
  selector: 'flight-flight-search',
  templateUrl: './flight-search.component.html',
  styleUrls: ['./flight-search.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FlightSearchComponent implements OnDestroy {
  private destroy$$ = new Subject<void>();

  searchFlightsFormGroup = new FormGroup({
    from: new FormControl(''),
    to: new FormControl(''),
  });

  flights$: Observable<Flight[]> =
    this.searchFlightsFormGroup.valueChanges.pipe(
      debounceTime(250),
      distinctUntilChanged(),
      switchMap((searchFlightsQueryParams) =>
        this.searchFlights(searchFlightsQueryParams as SearchFlightsQueryParams)
      ),
      takeUntil(this.destroy$$)
    );

  constructor(private flightService: FlightService) {}

  ngOnDestroy(): void {
    this.destroy$$.next();
  }

  private searchFlights(
    searchFlightsQueryParams: SearchFlightsQueryParams
  ): Observable<Flight[]> {
    return this.flightService.searchFlights(searchFlightsQueryParams);
  }
}
