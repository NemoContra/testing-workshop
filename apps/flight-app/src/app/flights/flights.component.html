<div class="card">
  <div class="header">
    <h2 class="title" data-tid="card-title">All Flights</h2>
  </div>
  <div class="content" data-tid="content">
    <div class="flight-cards" *ngIf="flights$ | async as flights">
      <flight-card
        *ngFor="let flight of flights"
        [flight]="flight"
      ></flight-card>
    </div>
  </div>
</div>
