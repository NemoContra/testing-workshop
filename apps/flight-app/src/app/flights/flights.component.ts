import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { FlightService } from '../services/flight.service';
import { Observable } from 'rxjs';
import { Flight } from '@testing-workshop/shared/util/api-interfaces';

@Component({
  selector: 'flight-flights',
  templateUrl: './flights.component.html',
  styleUrls: ['./flights.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class FlightsComponent implements OnInit {
  flights$?: Observable<Flight[]>;

  constructor(private flightService: FlightService) {}

  ngOnInit(): void {
    this.initFlights();
  }

  initFlights(): void {
    this.flights$ = this.flightService.getFlights();
  }
}
