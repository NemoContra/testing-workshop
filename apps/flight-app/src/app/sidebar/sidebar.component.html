<div class="sidebar-wrapper">
  <div class="logo">
    <a class="simple-text">
      <div class="logo-img">
        <img src="../../assets/img/angular2-logo.png" alt="" />
      </div>
      Flight42
    </a>
  </div>
  <ul class="nav">
    <li>
      <a routerLink="/" data-tid="link-home">
        <i class="ti-home"></i>
        <p>Home</p>
      </a>
    </li>

    <li>
      <a routerLink="/flights" data-tid="link-flights">
        <i class="ti-arrow-top-right"></i>
        <p>All Flights</p>
      </a>
    </li>

    <li>
      <a routerLink="/flight-search" data-tid="link-flight-search">
        <i class="ti-search"></i>
        <p>Search Flights</p>
      </a>
    </li>
  </ul>
</div>
