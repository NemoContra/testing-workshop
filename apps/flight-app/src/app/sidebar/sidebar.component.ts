import { Component } from '@angular/core';

@Component({
  selector: 'flight-sidebar',
  templateUrl: 'sidebar.component.html',
})
export class SidebarComponent {}
