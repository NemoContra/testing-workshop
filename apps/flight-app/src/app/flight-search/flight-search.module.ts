import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FlightSearchComponent } from './flight-search.component';
import { ReactiveFormsModule } from '@angular/forms';
import { ClientSharedUiFlightCardComponent } from '@testing-workshop/client-shared/ui/flight-card';

@NgModule({
  imports: [
    CommonModule,
    ReactiveFormsModule,
    ClientSharedUiFlightCardComponent,
  ],
  declarations: [FlightSearchComponent],
  exports: [FlightSearchComponent],
})
export class FlightSearchModule {}
