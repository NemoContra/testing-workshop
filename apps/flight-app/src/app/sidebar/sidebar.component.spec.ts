import { SidebarComponent } from './sidebar.component';
import { createRoutingFactory } from '@ngneat/spectator/jest';
import { ROUTES } from '../app-routing.module';
import { MockComponents } from 'ng-mocks';
import { HomeComponent } from '../home/home.component';
import { FlightsComponent } from '../flights/flights.component';
import { FlightSearchComponent } from '../flight-search/flight-search.component';
import { Location } from '@angular/common';
import { testIdSelector } from '@testing-workshop/shared/util/test-helpers';

describe('SidebarComponent', () => {
  const createComponent = createRoutingFactory({
    component: SidebarComponent,
    routes: ROUTES,
    declarations: [
      MockComponents(HomeComponent, FlightsComponent, FlightSearchComponent),
    ],
    stubsEnabled: false,
  });

  const selectors = {
    home: testIdSelector('link-home'),
    flights: testIdSelector('link-flights'),
    flightSearch: testIdSelector('link-flight-search'),
  } as const;

  it('should match snapshot', () => {
    const spectator = createComponent();
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should navigate home, flights and flight-search', () => {
    const spectator = createComponent();
    const location = spectator.inject(Location);

    spectator.click(selectors.home);
    expect(location.path()).toMatchInlineSnapshot(`"/"`);

    spectator.click(selectors.flights);
    expect(location.path()).toMatchInlineSnapshot(`"/flights"`);

    spectator.click(selectors.flightSearch);
    expect(location.path()).toMatchInlineSnapshot(`"/flight-search"`);
  });
});
