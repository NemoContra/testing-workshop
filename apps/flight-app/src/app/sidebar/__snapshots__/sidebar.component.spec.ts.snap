// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarComponent should match snapshot 1`] = `
<flight-sidebar
  __ngContext__="0"
>
  <div
    class="sidebar-wrapper"
  >
    <div
      class="logo"
    >
      <a
        class="simple-text"
      >
        <div
          class="logo-img"
        >
          <img
            alt=""
            src="../../assets/img/angular2-logo.png"
          />
        </div>
         Flight42 
      </a>
    </div>
    <ul
      class="nav"
    >
      <li>
        <a
          data-tid="link-home"
          href="/"
          routerlink="/"
        >
          <i
            class="ti-home"
          />
          <p>
            Home
          </p>
        </a>
      </li>
      <li>
        <a
          data-tid="link-flights"
          href="/flights"
          routerlink="/flights"
        >
          <i
            class="ti-arrow-top-right"
          />
          <p>
            All Flights
          </p>
        </a>
      </li>
      <li>
        <a
          data-tid="link-flight-search"
          href="/flight-search"
          routerlink="/flight-search"
        >
          <i
            class="ti-search"
          />
          <p>
            Search Flights
          </p>
        </a>
      </li>
    </ul>
  </div>
</flight-sidebar>
`;
